File: Assets/Python/root/processing/processor.py
import traceback
from threading import Thread
from typing import Generic, TypeVar, Any
from enum import IntEnum
import os

from processing.streams import Input, Output, Stream, SingleLabelList, MultiLabelList
from datastructs.datalist import DataList

class SpecialAction(IntEnum):
    NOTHING = 0
    TERMINATE = 1

T = TypeVar('T')
U = TypeVar('U')
class Processor(Generic[T, U]):
    def __init__(self, input: Input[U], output: Output[T]):
        self._thread = Thread(target=self.loop)
        self._terminated = False
        self._input = input 
        self._output = output
        self._threaded = True

    def process(self, t: T) -> U: 
        raise NotImplementedError("Process method was not implemented")

    def take(self) -> T: return self._output.output()
    def give(self, u: U): self._input.input(u)

    def on_exception(self, e: Exception):
        traceback.print_exc()
        os._exit(0)

    def should_terminate(self, a: Any) -> bool:
        return a is SpecialAction.TERMINATE
    
    def should_skip(self, a: Any) -> bool:
        return a is SpecialAction.NOTHING

    def loop(self):
        while True:
            try:   
                self.setup()
                break
            except Exception:
                print("Had an strange exception while setting up, trying again")
        while not self._terminated:
            try:
                item = self.take()
                if self.should_terminate(item):      
                    finish = self.terminate()              
                    if not finish: break
                p_item = self.process(item)             
                if self.should_skip(p_item): continue   
                self.give(p_item)
            except Exception as e:
                self.on_exception(e)
    
    def start(self):
        if self._threaded: self._thread.start()
        else: self.setup()

    def setup(self):
        pass

    def join(self, time) -> bool:
        if not self._threaded: return False
        self._thread.join(time)
        return self._thread.is_alive()

    def manual(self, t: T):
        if self._threaded: return
        self.give(self.process(t))

    def send_special(self, u: U):
        self.give(u)

    def terminate(self) -> bool: 
        self._terminated = True

    def external_terminate(self):
        if isinstance(self._output, Stream):
            self._output.input(SpecialAction.TERMINATE)

class NamedProcessor(Processor[T,U]):
    def __init__(self, name: str, input: Input[U], output: Output[T]):
        super().__init__(input, output)
        self.name = name

class DLInputter(NamedProcessor[DataList, U]):
    def should_terminate(self, dl: DataList) -> bool:
        if dl is None: return True
        return super().should_terminate(dl) or dl.has_terminal_command_for(self.name)
    
class DLOutputter(NamedProcessor[T, DataList]): pass
    
class DLProcessor(DLInputter[DataList]): 
    def process(self, dl: DataList) -> DataList:
        return dl
    
class RoutingDLProcessor(DLInputter[SingleLabelList]): pass
class LabellingDLProcessor(DLInputter[MultiLabelList]): pass

File: Assets/Scripts/processing/processor.cs
using System;
using System.Diagnostics;
using System.Threading;
using Debug = UnityEngine.Debug;
using System.Linq;
using System.Reflection;
using System.Collections.Generic;

public enum LoopMode {
    EVENT,
    THREADED,
    UNITY
}

public static class LoopModeExtension {
    public static bool IsBlocking(this LoopMode loopMode) {
        return loopMode == LoopMode.THREADED;
    }
}

public class SkipExeception : Exception {}
public class StopExeception : Exception {}

public interface IProcessor {
    void Start();
    bool Join(int msTimeout);
    void ExternalTerminate();
}

public abstract class Processor<T,U> : IProcessor {
    private Thread thread;
    public LoopMode loopMode;
    public IInput<U> input;
    public IOutput<T> output;
    protected bool terminated = false;

    public Processor(IInput<U> input, IOutput<T> output, LoopMode loopMode) {
        (this.input, this.output, this.loopMode) = (input, output, loopMode);
        if(loopMode == LoopMode.THREADED){
            thread = new(Loop);
        }
    }

    public bool IsTerminated() { return terminated; }
    public bool IsBlocking() { return loopMode.IsBlocking(); }

    public abstract U Process(T t);

    public virtual T Take() { return output.Output(IsBlocking()); }
    public virtual void Give(U u) { input.Input(u, IsBlocking()); }

    public virtual void OnExeception(Exception e) {
        Debug.Log(e.StackTrace);
        Debug.Break();
    }

    public void Loop() {
        while(!IsTerminated()) {
            Single(Process);
        }
        OnTermination();
    }

    public void Single(Func<T,U> func) {
        try {
            Give(func(Take()));
        } catch (SkipExeception) {
        } catch (StopExeception) {
            terminated = true;
        } catch (Exception e) {
            Debug.LogException(e);
            Debug.Break();
        }
    }

    public virtual void Start() {
        if(loopMode == LoopMode.THREADED) thread.Start();
    }

    public bool Join(int msTimeout) {
        if(loopMode != LoopMode.THREADED) return true;
        thread.Join(msTimeout);
        return !thread.IsAlive;
    }
    
    public virtual void InternalTerminate() {
        throw new StopExeception();
    }

    public virtual void ExternalTerminate() {
        terminated = true;
        FeedOutputNull();
    }

    public void FeedOutputNull() {
        (output as IInput<DataList>)?.Input(null, IsBlocking());
    }

    protected virtual void OnTermination() {}
}

public interface INameable {
    string GetName();
}

public abstract class NamedProcessor<T,U> : Processor<T,U>, INameable {
    public string name;

    public NamedProcessor(string name, IInput<U> input, IOutput<T> output, LoopMode loopMode) : base(input, output, loopMode) {
        this.name = name;
    }

    public string GetName() { return name; }
}

public abstract class DLInputter<U> : NamedProcessor<DataList, U> {
    protected DLInputter(string name, IInput<U> input, IOutput<DataList> output, LoopMode loopMode) : base(name, input, output, loopMode) {}

    public override DataList Take() {
        var dl = base.Take();
        if(dl == null || dl.HasTerminalCommand(name)) InternalTerminate();
        return dl;
    }
}

public abstract class DLOutputter<T> : NamedProcessor<T, DataList> {
    protected DLOutputter(string name, IInput<DataList> input, IOutput<T> output, LoopMode loopMode) : base(name, input, output, loopMode) {}
}

public abstract class DLProcessor : DLInputter<DataList> {
    protected DLProcessor(string name, IInput<DataList> input, IOutput<DataList> output, LoopMode loopMode) : base(name, input, output, loopMode) {}

    public override DataList Process(DataList dl) { return dl; }
}

public abstract class RoutingDLProcessor : DLInputter<SingleLabelList> {
    protected RoutingDLProcessor(string name, IInput<SingleLabelList> input, IOutput<DataList> output, LoopMode loopMode) : base(name, input, output, loopMode) {}
}

public abstract class LabellingDLProcessor : DLInputter<MultiLabelList> {
    protected LabellingDLProcessor(string name, IInput<MultiLabelList> input, IOutput<DataList> output, LoopMode loopMode) : base(name, input, output, loopMode) {}
}